src/devices/GoPiGo3/GoPiGo3.cs

```diff
@@ -0,0 +1,900 @@
+                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 };
+            byte[] reply = SpiTransferArray(outArray);
+
+            if (reply[3] == 0xA5)
+            {
+                for (int ndx = 4; ndx < 24; ++ndx)
+                {
+                    if (reply[ndx] != 0)
+                    {
+                        retVal += (char)(reply[ndx]);
+                    }
+                    else
+                    {
+                        break;
+                    }
+                }
+            }
+            else
+            {
+                throw new IOException("No SPI response");
+            }
+            return retVal;
+        }
+
+        /// <summary>
+        /// Read the hardware version
+        /// </summary>
+        /// <returns>Returns the hardware version as a string</returns>
+        public string GetHardwareVersion()
+        {
+            string retVal = string.Empty;
+            int version = SpiRead32(SpiMessageType.GetHardwareVersion);
+            retVal = $"{version / 1000000}.{(version / 1000) % 1000}.{version % 1000}";
+            return retVal;
+        }
+
+        /// <summary>
+        /// Read the 128 bit GoPiGo3 hardware serial number
+        /// </summary>
+        /// <returns>Returns the serial number as 32 char HEX formatted string</returns>
+        public string GetId()
+        {
+            string retVal = string.Empty;
+            byte[] outArray = {SpiAddress, (byte)SpiMessageType.GetId,
+                     0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 };
+            byte[] reply = SpiTransferArray(outArray);
+
+            if (reply[3] == 0xA5)
+            {
+                retVal = string.Join("", reply.Skip(4).Take(16).Select((b) => b.ToString("X2")));
+            }
+            else
+            {
+                throw new IOException("No SPI response");
+            }
+            return retVal;
+        }
+
+        /// <summary>
+        /// Read the firmware version
+        /// </summary>
+        /// <returns>Returns the firmware version</returns>
+        public string GetFirmwareVersion()
+        {
+            string retVal = string.Empty;
+            int version = SpiRead32(SpiMessageType.GetFirmwareVersion);
+            retVal = string.Format("{0}.{1}.{2}", (version / 1000000), ((version / 1000) % 1000), (version % 1000));
+            return retVal;
+        }
+
+        /// <summary>
+        /// Set the led color
+        /// </summary>
+        /// <param name="led">The led, either Left, Right, Blinky Left, Blinky Right and Wifi. Note you should only control the wifi one if you are sure to be connected to wifi</param>
+        /// <param name="red">The red component</param>
+        /// <param name="green">The green component</param>
+        /// <param name="blue">The blue component</param>
+        public void SetLed(byte led, byte red, byte green = 0, byte blue = 0)
```

Comment:
wondering if should use [Color](https://docs.microsoft.com/en-us/dotnet/api/system.drawing.color?view=netcore-2.1) instead of RGB

Reply:
yes, implementing as well